<a href="https://colab.research.google.com/github/ggoddll99/24-2ESAA/blob/main/241004_%EA%B3%BC%EC%A0%9C_%EB%AC%B8%EC%84%9C%EA%B5%B0%EC%A7%91%ED%99%94.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **07. 문서 군집화 소개와 실습**

* 텍스트 군집화, 유사도 측정 : 비슷한 유형의 문서에 대해 군집화를 수행하는 기법. 텍스트 분류를 비지도학습으로 수행하는 방법.

* 텍스트 분류 : 문서가 특정 분류 또는 카테고리에 속하는 것을 예측하는 기법. - ex. 특정 신문 기사 내용이 연예/정치/사회/문화 중 어떤 카테고리에 속하는지 자동으로 분류하거나 스팸 메일 검출 같은 프로그램이 이에 속함. 지도학습 적용

텍스트 분류 기반의 문서 분류는 사전에 결정 카테고리 값을 가진 학습 데이터 세트가 필요한데, 문서 군집화는 학습 데이터 세트가 필요 없는 비지도학습 기반으로 동작한다.

Opinion Review 데이터 세트 : 51개의 텍스트 파일로 구성돼 있으며, 각 파일은 Tripadvisor(호텔), Edmunds.com(자동차), Amazon.com(전자제품) 사이트에서 가져온 리뷰 문서이다. 각 문서는 약 100개 정도의 문장을 가지고 있다. 원래는 토픽 모델링 논문으로 사용된 데이터 세트인데, 여기서는 문서 군집화를 이용해 각 리뷰를 분류한다.

https://archive.ics.uci.edu/dataset/191/opinosis+opinion+frasl+review

압축을 풀면 디렉터리로 되어있으며, topics 디렉터리 안에 51개의 파일로 구성되어 있다.

accuracy_garmin_nuvi_255W_gps.txt.data는 Amazon.com에서 Garmin nuvi라는 차량용 내비게이션 모델에 대한 리뷰를 담고 있다.

해당 디렉터리 내에 있는 파일을 하나씩 읽어서 파일명과 파일 리뷰를 하나의 DataFrame으로 로드하여 파일명별로 어떤 리뷰를 담고 있는지 개략적으로 살펴본다.

이번 예제는 데이터를 로드할 때 절대 경로 디렉터리를 먼저 지정한 뒤, 이 위치에서 데이터를 로딩한다.

**데이터 로딩 방법**
1. 해당 디렉터리 내의 모든 파일에 대해 각각 for 반복문으로 반복하면서 개별 파일명을 파일명 리스트에 추가하고 개별 파일은 DataFrame으로 읽은 후 다시 문자열로 반환한 뒤 파일 내용 리스트에 추가한다.
2. 만들어진 파일명 리스트와 파일 내용 리스트를 이용해 새롭게 파일명과 파일 내용을 칼럼으로 가지는 DataFrame을 생성한다.

In [1]:
import pandas as pd
import glob, os
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth', 700)

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
path=r'/content/drive/MyDrive/24-2/ESAA/과제/OpinosisDataset1.0/topics'
# path로 지정한 디렉터리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
all_files=glob.glob(os.path.join(path, '*.data'))
filename_list=[]
opinion_text=[]

# 개별 파일들의 파일명은 filename_list 리스트로 취합
# 개별 파일들의 파일 내용은 DataFrame 로딩 후 다시 string으로 변환하여 opinion_text 리스트로 취합
for file_ in all_files:
  # 개별 파일을 읽어서 DataFrame으로 생성
  df=pd.read_table(file_, index_col=None, header=0, encoding='latin1')
  # 절대경로로 주어진 파일명을 가공
  # 맨 마지막 .data 확장자 제거
  filename_=file_.split('/')[-1]
  filename=filename_.split('.')[0]

  # 파일명 리스트와 파일 내용 리스트에 파일명과 파일 내용 추가
  filename_list.append(filename)
  opinion_text.append(df.to_string())

# 파일명 리스트와 파일 내용 리스트를 DataFrame으로 생성
document_df=pd.DataFrame({'filename' : filename_list, 'opinion_text' : opinion_text})
document_df.head()

filename  \
0  accuracy_garmin_nuvi_255W_gps   
1      comfort_honda_accord_2008   
2   display_garmin_nuvi_255W_gps   
3     battery-life_amazon_kindle   
4          buttons_amazon_kindle   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  
0                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...  
1                                                                                                                                                                                                                                                                                                                                                                                                Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...  
2                                                                                                                                                                                                                                                       3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                                                                                                         ...  
3                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...  
4                                                                                                                                             I thought it would be fitting to christen my Kindle with the Stephen King novella UR, so went to the Amazon site on my computer and clicked 

3. 문서를 TF-IDF 형태로 피처 벡터화한다. `TfidfVectorizer`는 Lemmatization 같은 어근 변환을 직접 지원하진 않지만, `tokenizer` 인자에 커스텀 어근 변환 함수를 적용해 어근 변환을 수행할 수 있다. 이를 위해 `LemNormalize()` 함수를 만든다.
  * `ngram` (1, 2)
  * `min_df`
  * `max_df`

`TfidfVectorizer`의 `fit_transform()`의 인자로 `document_df` DataFrame의 `opinion_text` 칼럼을 입력하면 개별 문서 텍스트에 대해 TF-IDF 변환된 피처 벡터화된 행렬을 구할 수 있다.

In [6]:
from nltk.stem import WordNetLemmatizer
import nltk
nltk.download('punkt')
nltk.download('wordnet')
import string

remove_punct_dict=dict((ord(punct), None) for punct in string.punctuation)
lemmar=WordNetLemmatizer()

def LemTokens(tokens):
  return [lemmar.lemmatize(token) for token in tokens]

def LemNormalize(text):
  return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


In [7]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect=TfidfVectorizer(tokenizer=LemNormalize, stop_words='english',
                           ngram_range=(1, 2), min_df=0.05, max_df=0.85)

# opinion_text 칼럼값으로 feature vectorization 수행
feature_vect=tfidf_vect.fit_transform(document_df['opinion_text'])

문서별 텍스트가 TF-IDF 변환된 피처 벡터화 행렬 데이터에 대해서 군집화를 수행해 어떤 문서끼리 군집되는지 확인해 보자. 군집화 기법은 K-평균을 적용한다. 문서의 유형은 크게 보자면, 전자제품, 자동차, 호텔로 돼 있다.

4. 먼저 5개의 중심 기반으로 어떻게 군집화되는지 확인한다.
  * 최대 반복 횟수 `max_iter` 10000

5. KMeans 수행 후 군집의 Label 값과 중심별로 할당된 데이터 세트의 좌표 값을 구한다.

In [8]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행. 예제를 위해 동일한 클러스터링 결과 도출용 random_state=0
km_cluster=KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label=km_cluster.labels_
cluster_centers=km_cluster.cluster_centers_

6. 각 데이터별로 할당된 군집의 레이블을 파일명과 파일 내용을 가지고 있는 `document_df` DataFrame에 'cluster_label' 칼럼을 추가해 저장한다. 각 파일명은 의견 리뷰에 대한 주제를 나타낸다. 군집이 각 주제별로 유사한 형태로 잘 구성됐는지 알아보자.

In [9]:
document_df['cluster_label']=cluster_label
document_df.head()

filename  \
0  accuracy_garmin_nuvi_255W_gps   
1      comfort_honda_accord_2008   
2   display_garmin_nuvi_255W_gps   
3     battery-life_amazon_kindle   
4          buttons_amazon_kindle   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  \
0                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
1                                                                                                                                                                                                                                                                                                                                                                                                Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
2                                                                                                                                                                                                                                                       3 quot  widescreen display was a bonus .\n0                                                                                                                                           This made for smoother graphics on the 255w of the vehicle moving along displayed roads, where the 750's display was more of a  jerky  movement .\n1                                                                                                                         ...   
3                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
4                                                                                                                                             I thought it would be fitting to christen my Kindle with the Stephen King novella UR, so went to the Amazon site on my computer and cli

판다스 DataFrame의 `sort_values(by=정렬칼럼명)`를 수행하면 인자로 입력된 '정렬칼럼명'으로 데이터를 정렬할 수 있다.

7. document_df DataFrame 객체에서 cluster_label로 어떤 파일명으로 매칭됐는지 보면서 군집화 결과를 확인한다.

In [10]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
7   bathroom_bestwestern_hotel_sfo   
16         food_holiday_inn_london   
24          food_swissotel_chicago   
22      free_bestwestern_hotel_sfo   
25  location_bestwestern_hotel_sfo   
18     location_holiday_inn_london   
20   parking_bestwestern_hotel_sfo   
29             price_amazon_kindle   
27        price_holiday_inn_london   
36         room_holiday_inn_london   
42     rooms_bestwestern_hotel_sfo   
35         rooms_swissotel_chicago   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
7                                                                                                                                                                                                                                                 The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
16                                                                                                                                                                                                                                                                      The room was packed to capacity with queues at the food buffets .\n0                                                                                                                                                   The over zealous   staff cleared our unfinished drinks while we were collecting cooked food and movement around the room with plates was difficult in the crowded circumstances .\n1                                         ...   
24                                                                                                                                                                                                                                                                                                                                                                   The food for our event was delicious .\n0                                                                                                                                                                                                                                                                                                              ...   
22                                                                                                                                                                The wine reception is a great idea as it is nice to meet other travellers and great having access to the free Internet access in our room .\n0                                                                                      They also have a computer available with free internet which is a nice bonus but I didn't find that out till the day before we left but was still able to get on there to check our flight to Vegas the next day .\n1                                                                                                 ...

Cluster #0은 호텔

In [11]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
0     accuracy_garmin_nuvi_255W_gps   
3        battery-life_amazon_kindle   
9        battery-life_ipod_nano_8gb   
6       battery-life_netbook_1005ha   
4             buttons_amazon_kindle   
8   directions_garmin_nuvi_255W_gps   
2      display_garmin_nuvi_255W_gps   
14    eyesight-issues_amazon_kindle   
15              fonts_amazon_kindle   
17          keyboard_netbook_1005ha   
13         navigation_amazon_kindle   
32   satellite_garmin_nuvi_255W_gps   
30      screen_garmin_nuvi_255W_gps   
28             screen_ipod_nano_8gb   
33            screen_netbook_1005ha   
41         size_asus_netbook_1005ha   
40       speed_garmin_nuvi_255W_gps   
46     updates_garmin_nuvi_255W_gps   
45       voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
0                                                                                                                                                                                                                                                                  , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
3                                                                                                                                                                                                                After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
9                                                                                                                                                                                                                                                                                                  short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
6                                                                                                                                                                                                                                                                                                                   6GHz 533FSB cpu, glossy display, 3, Cell 23Wh Li, ion Battery  , and a 1 .\n0                                 

Cluster #1 : 네비/킨들, 아이팟, 넷북

In [14]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
19    features_windows7   
37  sound_ipod_nano_8gb   
43       speed_windows7   
50  video_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
19                                                                                                                                                                             I had to uninstall anti, virus and selected other programs, some of which did not have listings in the  Programs and Features  Control Panel section .\n0                                                                                                                                                                                                                           This review briefly touches upon some of the key features and enhancements of Microsoft's latest OS .\n1                                                 ...   
37     headphone jack i got a clear case for it and it  i got a clear case for it and it like prvents me from being able to put the jack all the way in so the sound can b messsed up or i can get it in there and its playing well them go to move or something and it slides out .\n0                                                                                                                                                                                                                Picture and sound quality are excellent for this typ of devic .\n1                                                                                                                                                   ...   
43                                                                                                                        Windows 7 is quite simply faster, more stable, boots faster, goes to sleep faster, comes back from sleep faster, manages your files better and on top of that it's beautiful to look at and easy to use .\n0                                                                                                                                                                                                                                           , faster about 20% to 30% faster at running applications than my Vista ,  seriously\n1                                                     ...   
50                                                                                                                                                                                                                                                                           I bought the 8, gig Ipod Nano that has the built, in video camera .\n0                                                                                                                                                                                                                      Itunes has an on, line store, where you may purchase and download music and videos which will install onto the ipod .\n1                           ...   

    cluster_label  
19              2  
37              2  
43              2  
50              2

Cluster #2 : 인터넷, 아이팟

In [13]:
document_df[document_df['cluster_label']==3].sort_values(by='filename')

filename  \
31    service_bestwestern_hotel_sfo   
39       service_holiday_inn_london   
44  service_swissotel_hotel_chicago   
48      staff_bestwestern_hotel_sfo   
49          staff_swissotel_chicago   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
31                                                                                                                                                           Both of us having worked in tourism for over 14 years were very disappointed at the level of service provided by this gentleman .\n0                                                                                                                                                                                              The service was good, very friendly staff and we loved the free wine reception each night .\n1                                                                                                                               ...   
39                                                                                                                                                                                                                                                                                                      not customer, oriented hotelvery low service levelboor reception\n0                                                                                                                                                                                                                                                                    The room was quiet, clean, the bed and pillows were comfortable, and the serv...   
44                                                                                                                                                                                                                                                                                                                                                  Mediocre room and service for a very extravagant price .\n0                                                                                                                                                                                                                                                                                                             ...   
48                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Staff are friendl...   
49                                                                                                                                                                                                                                                               

Cluster #3 : 호텔

In [15]:
document_df[document_df['cluster_label']==4].sort_values(by='filename')

filename  \
1        comfort_honda_accord_2008   
5        comfort_toyota_camry_2007   
21   gas_mileage_toyota_camry_2007   
11      interior_honda_accord_2008   
10      interior_toyota_camry_2007   
12       mileage_honda_accord_2008   
23   performance_honda_accord_2008   
38      performance_netbook_1005ha   
26       quality_toyota_camry_2007   
34         seats_honda_accord_2008   
47  transmission_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
1                                                                                                                                                                                                                                                                                                                                                                                                 Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
5                                                                                                                                                   Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                        Seats are fine, in fact of all the smaller sedans this is the most comfortable I found for the price as I am 6', 2  and 250# .\n1                                                                                                                                                                                     Great gas mileage and comfortable on long trips ...   
21                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                                                                                                                                                  ...   
11                                                                                                                                                                                                                                                                                                                                                                                                                                                                             I love the new body style and the interior is a simple pleasure except for the center dash .\n0                                                                                                                                              ...   
10                                

Cluster #4 : 자동차

전반적으로 군집화된 결과를 살펴보면 군집 개수가 약간 많게 설정되어 있어서 세분화되어 군집화되었다. 중심 개수를 5개에서 3개로 낮춰서 3개 그룹으로 군집화해보자.

In [16]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화
km_cluster=KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label=km_cluster.labels_

# 소속 클러스터를 cluster_label 칼럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label']=cluster_label
document_df.sort_values(by='cluster_label')

filename  \
25   location_bestwestern_hotel_sfo   
22       free_bestwestern_hotel_sfo   
27         price_holiday_inn_london   
20    parking_bestwestern_hotel_sfo   
18      location_holiday_inn_london   
16          food_holiday_inn_london   
31    service_bestwestern_hotel_sfo   
49          staff_swissotel_chicago   
35          rooms_swissotel_chicago   
36          room_holiday_inn_london   
24           food_swissotel_chicago   
7    bathroom_bestwestern_hotel_sfo   
42      rooms_bestwestern_hotel_sfo   
44  service_swissotel_hotel_chicago   
48      staff_bestwestern_hotel_sfo   
39       service_holiday_inn_london   
40       speed_garmin_nuvi_255W_gps   
32   satellite_garmin_nuvi_255W_gps   
41         size_asus_netbook_1005ha   
30      screen_garmin_nuvi_255W_gps   
29              price_amazon_kindle   
28             screen_ipod_nano_8gb   
45       voice_garmin_nuvi_255W_gps   
46     updates_garmin_nuvi_255W_gps   
33            screen_netbook_1005ha   
0     accuracy_garmin_nuvi_255W_gps   
17          keyboard_netbook_1005ha   
15              fonts_amazon_kindle   
14    eyesight-issues_amazon_kindle   
13         navigation_amazon_kindle   
2      display_garmin_nuvi_255W_gps   
3        battery-life_amazon_kindle   
4             buttons_amazon_kindle   
6       battery-life_netbook_1005ha   
9        battery-life_ipod_nano_8gb   
8   directions_garmin_nuvi_255W_gps   
26        quality_toyota_camry_2007   
1         comfort_honda_accord_2008   
47   transmission_toyota_camry_2007   
43                   speed_windows7   
5         comfort_toyota_camry_2007   
38       performance_netbook_1005ha   
37              sound_ipod_nano_8gb   
10       interior_toyota_camry_2007   
11       interior_honda_accord_2008   
34          seats_honda_accord_2008   
12        mileage_honda_accord_2008   
19                features_windows7   
21    gas_mileage_toyota_camry_2007   
23    performance_honda_accord_2008   
50              video_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
25                                                                                                                                                                                                                                                                                                       Good Value good location ,  ideal choice .\n0                                                                                                                                                                                                                                                                                          Great Location ,  Nice   Rooms ,  Helpless Concierge\n1                     ...   
22                                                                                                                                                                The wine reception is a great idea as it is nice to meet other travellers and great having access to the free Internet access in our room .\n0                                                                                      They also have a computer available with free internet which is a nice bonus but I didn't find that out till the day before we left but was still able to get on there to check our flight to Vega

In [17]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
7    bathroom_bestwestern_hotel_sfo   
16          food_holiday_inn_london   
24           food_swissotel_chicago   
22       free_bestwestern_hotel_sfo   
25   location_bestwestern_hotel_sfo   
18      location_holiday_inn_london   
20    parking_bestwestern_hotel_sfo   
27         price_holiday_inn_london   
36          room_holiday_inn_london   
42      rooms_bestwestern_hotel_sfo   
35          rooms_swissotel_chicago   
31    service_bestwestern_hotel_sfo   
39       service_holiday_inn_london   
44  service_swissotel_hotel_chicago   
48      staff_bestwestern_hotel_sfo   
49          staff_swissotel_chicago   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
7                                                                                                                                                                                                                                                 The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
16                                                                                                                                                                                                                                                                      The room was packed to capacity with queues at the food buffets .\n0                                                                                                                                                   The over zealous   staff cleared our unfinished drinks while we were collecting cooked food and movement around the room with plates was difficult in the crowded circumstances .\n1                                         ...   
24                                                                                                                                                                                                                                                                                                                                                                   The food for our event was delicious .\n0                                                                                                                                                                                                                                                                                                              ...   
22                                                                                                                                                                The wine reception is a great idea as it is nice to meet other travellers and great having access to the free Internet access in our room .\n0                                                                                      They also have a computer available with free internet which is a nice bonus but I didn't find that out till the day before we left but was still

Cluster #0 : 호텔

In [18]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
0     accuracy_garmin_nuvi_255W_gps   
3        battery-life_amazon_kindle   
9        battery-life_ipod_nano_8gb   
6       battery-life_netbook_1005ha   
4             buttons_amazon_kindle   
8   directions_garmin_nuvi_255W_gps   
2      display_garmin_nuvi_255W_gps   
14    eyesight-issues_amazon_kindle   
15              fonts_amazon_kindle   
17          keyboard_netbook_1005ha   
13         navigation_amazon_kindle   
29              price_amazon_kindle   
32   satellite_garmin_nuvi_255W_gps   
30      screen_garmin_nuvi_255W_gps   
28             screen_ipod_nano_8gb   
33            screen_netbook_1005ha   
41         size_asus_netbook_1005ha   
40       speed_garmin_nuvi_255W_gps   
46     updates_garmin_nuvi_255W_gps   
45       voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
0                                                                                                                                                                                                                                                                  , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
3                                                                                                                                                                                                                After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
9                                                                                                                                                                                                                                                                                                  short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
6                                                                                                                                                                                                                                                                                                                   6GHz 533FSB cpu, glossy display, 3, Cell 23Wh Li, ion Battery  , and a 

Cluster #1 : 포터블 기기(네비, 넷북, 킨들)

In [19]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
1        comfort_honda_accord_2008   
5        comfort_toyota_camry_2007   
19               features_windows7   
21   gas_mileage_toyota_camry_2007   
11      interior_honda_accord_2008   
10      interior_toyota_camry_2007   
12       mileage_honda_accord_2008   
23   performance_honda_accord_2008   
38      performance_netbook_1005ha   
26       quality_toyota_camry_2007   
34         seats_honda_accord_2008   
37             sound_ipod_nano_8gb   
43                  speed_windows7   
47  transmission_toyota_camry_2007   
50             video_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
1                                                                                                                                                                                                                                                                                                                                                                                                 Drivers seat not comfortable, the car itself compared to other models of similar class .\n0                                                                                                                                                                                                                               ...   
5                                                                                                                                                   Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                        Seats are fine, in fact of all the smaller sedans this is the most comfortable I found for the price as I am 6', 2  and 250# .\n1                                                                                                                                                                                     Great gas mileage and comfortable on long trips ...   
19                                                                                                                                                                             I had to uninstall anti, virus and selected other programs, some of which did not have listings in the  Programs and Features  Control Panel section .\n0                                                                                                                                                                                                                           This review briefly touches upon some of the key features and enhancements of Microsoft's latest OS .\n1                                                 ...   
21                                                                                                                                                                                                                                                                                                                                                                              Ride seems comfortable and gas mileage fairly good averaging 26 city and 30 open road .\n0                                                                                                                                   

Cluster #2 : 인터넷, 아이팟, 자동차(다수), 넷북

**군집별 핵심 단어 추출**

각 군집(Cluster)에 속한 문서는 핵십 단어를 주축으로 군집화돼 있을 것이다. 각 군집을 구성하는 핵심 단어가 어떤 것이 있는지 확인해 보자.

KMeans 객체는 각 군집을 구성하는 단어 피처가 군집의 중심을 기준으로 얼마나 가깝게 위치해 있는지 `cluster_centers_`라는 속성으로 제공한다. `cluster_centers_`는 배열 값으로 제공되며, 행은 개별 군집을, 열은 개별 피처를 의미한다. 각 배열 내의 값은 개별 군집 내의 상대 위치를 숫자 값으로 표현한 일종의 좌표 값이다. 예를 들어 `cluster_centers[0, 1]`은 0번 군집에서 두 번째 피처의 위치 값이다. 바로 앞 예제에서 군집 3개로 생성한 KMeans 객체인 `km_cluster`에서 `cluster_centers_` 속성 값을 가져온 뒤 값을 확인해 보자.

In [20]:
cluster_centers=km_cluster.cluster_centers_
print('cluster_centers shape : ', cluster_centers.shape)
print(cluster_centers)

cluster_centers shape :  (3, 4611)
[[0.         0.00099499 0.00174637 ... 0.         0.00183397 0.00144581]
 [0.00726575 0.         0.         ... 0.00882859 0.         0.        ]
 [0.0070677  0.00061701 0.         ... 0.         0.         0.        ]]


* 군집 3개, word 피처 4611개

각 행의 배열 값은 각 군집 내의 4611개 피처의 위치가 개별 중심과 얼마나 가까운가를 상대 값으로 나타낸 것이다. 0에서 1까지의 값을 가질 수 있으며 1에 가까울수록 중심과 가까운 값을 의미한다.

`cluster_centers_` 속성값을 이용해 각 군집별 핵심 단어를 찾아보자. `cluster_centers_` 속성은 넘파이의 ndarray이다. ndarray의 `argsort()[:, ::-1]`를 이용하면 `cluster_centers` 배열 내 값이 큰 순으로 정렬된 위치 인덱스 값을 반환한다(큰 값으로 정렬한 값을 반환하는 게 아니라 큰 값을 가진 배열 내 위치 **인덱스 값**을 반환하는 것이다). 이 위치 인덱스 값이 필요한 이유는 핵심 단어 피처의 이름을 출력하기 위해서이다.

함수 `get_cluster_details()`는 `cluster_centers_` 배열 내에서 가장 값이 큰 데이터의 위치 인덱스를 추출한 뒤, 해당 인덱스를 이용해 핵심 단어 이름과 그때의 상대 위치 값을 추출해 `cluster_details`라는 Dict 객체 변수에 기록하고 반환한다.

In [33]:
# 군집별 top n 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명을 반환함.
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
  cluster_details={}

  # cluster_centers array의 값이 큰 순으로 정렬된 인덱스 값 반환
  # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함.
  centroid_feature_ordered_ind=cluster_model.cluster_centers_.argsort()[:, ::-1]

  # 개별 군집별로 반복하면서 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명 입력
  for cluster_num in range(clusters_num):
    # 개별 군집별 정보를 담을 데이터 초기화
    cluster_details[cluster_num]={}
    cluster_details[cluster_num]['cluster']=cluster_num

    # cluster_centers_argsort()[:, ::-1]로 구한 인덱스를 이용해 top n 피처 단어를 구함
    top_feature_indexes=centroid_feature_ordered_ind[cluster_num, :top_n_features]
    top_features=[feature_names[ind] for ind in top_feature_indexes]

    # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함.
    top_feature_values=cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()

    # cluster_details 딕셔너리 객체에 개별 군집별 핵심단어와 중심위치 상댓값, 해당 파일명 입력
    cluster_details[cluster_num]['top_features']=top_features
    cluster_details[cluster_num]['top_features_value']=top_feature_values
    filenames=cluster_data[cluster_data['cluster_label']==cluster_num]['filename']
    filenames=filenames.values.tolist()

    cluster_details[cluster_num]['filenames']=filenames

  return cluster_details

`get_cluster_details()`를 호출하면 dictionary를 원소로 가지는 리스트인 `cluster_details`를 반환한다. `cluster_details`에는 개별 군집번호, 핵심 단어, 핵심단어 중심 위치 상댓값, 파일명 속성 값 정보가 있는데, 좀 더 보기 좋게 표현하기 위해 별도의 `print_cluster_details()` 함수를 만든다.

In [22]:
def print_cluster_details(cluster_details):
  for cluster_num, cluster_detail in cluster_details.items():
    print('####### Cluster {0}'.format(cluster_num))
    print('Top features:', cluster_detail['top_features'])
    print('Reviews 파일명 : ', cluster_detail['filenames'][:7])
    print('=================================================')

`get_cluster_details()`의 인자는 KMeans 군집화 객체, 파일명 추출을 위한 `document_df` DataFrame, 핵심 단어 추출을 위한 피처명 리스트, 전체 군집 개수, 핵심 단어 추출 개수이다. 피처명 리스트는 앞에서 TF-IDF 변환된 `tfidf_vect` 객체에서 `get_feature_names()`로 추출한다.

In [34]:
feature_names=tfidf_vect.get_feature_names_out()

cluster_details=get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,
                                    feature_names=feature_names, clusters_num=3, top_n_features=10)
print_cluster_details(cluster_details)

####### Cluster 0
Top features: ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 :  ['bathroom_bestwestern_hotel_sfo', 'food_holiday_inn_london', 'location_holiday_inn_london', 'parking_bestwestern_hotel_sfo', 'free_bestwestern_hotel_sfo', 'food_swissotel_chicago', 'location_bestwestern_hotel_sfo']
####### Cluster 1
Top features: ['screen', 'battery', 'keyboard', 'kindle', 'battery life', 'direction', 'life', 'voice', 'map', 'size']
Reviews 파일명 :  ['accuracy_garmin_nuvi_255W_gps', 'display_garmin_nuvi_255W_gps', 'battery-life_amazon_kindle', 'buttons_amazon_kindle', 'battery-life_netbook_1005ha', 'directions_garmin_nuvi_255W_gps', 'battery-life_ipod_nano_8gb']
####### Cluster 2
Top features: ['interior', 'seat', 'performance', 'mileage', 'comfortable', 'gas', 'quality', 'gas mileage', 'transmission', 'car']
Reviews 파일명 :  ['comfort_honda_accord_2008', 'comfort_toyota_camry_2007', 'interior_toyota_camry_2007', 'interior_honda

Cluster #0 : 호텔 리뷰 군집이기 때문에 'room' 'hotel', 'service', 'staff' 등과 같은 방과 서비스 등이 핵심 단어로 군집화되었다. 호텔 리뷰의 경우 방의 크기나 청소 상태, 직원들의 서비스, 위치 등이 주요 관심사이다.

Cluster #1 : 'screen', 'battery', 'battery life', 'keyboard' 등과 같은 화면과 배터리 수명 등이 핵심 단어로 군집화되었다. 모바일형이고 엔터테인먼트용 전자제품의 경우 화면 크기와 배터리 수명이 주요 관심사인듯 보인다.

Cluster #2 : 'interior', 'seat', 'mileage', 'comfortable' 등과 같은 실내 인테리어, 좌석, 연료 효율, 편안함 등이 핵심 단어로 군집화되었다. 토요타, 혼다와 같은 일본 자동차의 경우 실내 인테리어와 연료 효율, 편안함이 주요 관심사로 보인다.